files/public.go

```diff
@@ -119,6 +119,28 @@ func (storage *PublishedStorage) RemoveDirs(path string, progress aptly.Progress
 	return os.RemoveAll(filepath)
 }
 
+// Cleans up PublishedStorage tree towards the root up to first non-empty directory
+func (storage *PublishedStorage) CleanupPublishTree(path string) {
+	/*
+		Say, `path` is "a/b/c/d", let's remove all empty dirs:
+		 storage.rootPath + "a/b/c/d"
+		 storage.rootPath + "a/b/c"
+		 storage.rootPath + "a/b"
+		 ... and so on, up to first non-empty directory
+	*/
+	segments := strings.Split(path, "/")
+
+	for segmentsCount := len(segments); segmentsCount > 0; segmentsCount-- {
+		err := os.Remove(
+			filepath.Join(storage.rootPath,
+				strings.Join(segments[:segmentsCount], "/")))
+
+		if err != nil && strings.Contains(err.Error(), "directory not empty") {
```

Comment:
could we rather check if it is empty ?

Reply:
We could, in theory, but it is a lot more code, than checking an error...

Reply:
i know, but it is also more proper, to not accidentally delete stuff :)

Reply:
I do not agree with you. Also, it is algorithmically more complex to perform the check you propose.

In general, the question is: why do you trust one syscall and do not trust another?

Reply:
I think it is a question of taste... ;-) in some industries it is even not allowed to rely on error text, ... not a question of trusting sys calls..

in aptly I would prefer the check before action, we could add a function to utils/utils.go:
```
func isDirEmpty(path string) (bool, error) {
	f, err := os.Open(path)
	if err != nil {
		return false, err
	}
	defer f.Close()

	contents, err := f.Readdirnames(1) // Read only 1 file to check if it's empty
	if err != nil {
		return false, err
	}

	return len(contents) == 0, nil
}
```

and just call this before the os.Remove call. I would not worry about algorithic complexity with a low number of nested directories..

what do you think ?

Reply:
Look, even if os.Remove broke and start removing non-empty directories, it won't go beyond publish root. Also, if any of publishes are accidentally lost (which I highly doubt), they could be recreated from corresponding repos.

Anyway, I'm out of office for today and do not have a laptop at hand. Let's get back to it tomorrow.

Reply:
Also, if the root of your comment is that it is no good checking error **text**, we could simply bail out on any error. It is perfectly fine also.

Reply:
Look, the file system might also be on a NFS server, and I would like aptly to check before a remove action and have syscalls return errors on real errors. Yes it works both ways, but here in aptly i feel we should do it this way ;-)

Aside from this, there are some tests failing, maybe the tests were wrong, relying on the empty directory ?

Reply:
While I'm dealing with failed tests, let's make final round on os.Remove )

I, for one thrive to make code that from one side is bulletproof but from the other is not overloaded with useless actions and expressive enough to read as a poem.

It takes knowing the underlying api well, though.

As for os.Remove, it is wrapper for syscall.Rmdir in case we're removing a directory. Let's dive into various syscall docs now. Follow me.

* [Windows](https://github.com/golang/go/blob/master/src/syscall/syscall_windows.go#L580): it's a [RemoveDirectoryA](https://learn.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-removedirectorya#return-value) function, which docs say, it "Deletes an existing empty directory."
* [freebsd](https://github.com/golang/go/blob/master/src/syscall/zsyscall_freebsd_amd64.go#L949): it's a rmdir syscall, which [docs say](https://man.freebsd.org/cgi/man.cgi?query=rmdir&apropos=0&sektion=2&manpath=FreeBSD+14.1-RELEASE+and+Ports&arch=default&format=html) "The directory must not have any entries other than `.' and `..'."
* [linux](https://github.com/golang/go/blob/master/src/syscall/syscall_linux.go#L322): https://linux.die.net/man/2/unlinkat -> https://linux.die.net/man/2/rmdir
* [darwin](https://github.com/golang/go/blob/master/src/syscall/zsyscall_darwin_amd64.go#L1340): http://www.polarhome.com/service/man/?qf=RMDIR&af=0&sf=2&of=Darwin&tf=2
* [WASM](https://github.com/golang/go/blob/master/src/syscall/fs_wasip1.go#L697): https://wasix.org/docs/api-reference/wasi/path_remove_directory (read the note at the bottom)

All of them tell the same: it removes an empty directory, returns error otherwise. It's quite normal to try and remove non-empty  directory, handle error, move on. That's what is happening in my code. We're making a syscall and handling the error.

You say: 
>  i feel we should do it this way

Gut feeling is always okay, but you should not follow it blindly!

I strongly believe that review process is a place where we can learn something new and improve our skills every day - that's what is happening here as we type )

If you are still hesitant on the patch set, let's add more reviewers. And be sure, we had internal review, too, and our team came to the conclusion, that checking directory emptiness before os.Remove is excessive, more over, it creates additional points of failure.

Reply:
the expectation is that aptly logs if something unexpected happens. and Remove might fail due to various reasons, and we cannot rey on error messages. the only proper way is to check before, bail out if not empty, and log the proper os.Remove error if any, instead of relying on a error to happen. please respect that ;-)

Reply:
With all due respect, your approach is error-prone.

https://stackoverflow.com/a/78907172

As said, in Windows actual file removal may be deferred (as stated [in the docs](https://learn.microsoft.com/en-us/windows/win32/api/winbase/nf-winbase-deletefile)), so your userspace check might falsely fail due to that. On the other hand, operations performed in the OS kernel are well aware about all cached modifications made to fs.

As I already stated, feel free adding your implementation with separate commit then.